File: 1 Python/solutions/tictactoe.py
'''
Tic tac toe
'''

class Player:
    '''
    Player model
    '''
    def __init__(self, name, token):
        self.name = name
        self.token = token

    def __repr__(self):
        return self.token


class TicTacToe:
    '''
    Tic tac toe game
    '''
    def __init__(self):
        '''
        represent board as 3x3 list
        '''
        self.board = [[' ' for i in range(3)] 
                           for i in range(3)]

    def __repr__(self):
        '''
        formatted string representation of board
        '''
        draw_board = ''
        for line in self.board:
            line = [str(token) for token in line]
            draw_board += '|'.join(line) + '\n'
        return draw_board

    def move(self, x, y, player):
        '''
        moves player onto board if the position is unoccupied
        '''
        if type(self.board[y][x]) is Player: # position is taken
            return False
        self.board[y][x] = player
        return True

    def calc_winner(self):
        '''
        returns winner if one exists or none
        '''
        for i in range(3):
            # horizontal win
            if type(self.board[i][0]) is Player and \
                (self.board[i][0] == self.board[i][1] == self.board[i][2]):
                    return self.board[i][0]
            # vertical win
            if type(self.board[0][i]) is Player and \
                (self.board[0][i] == self.board[1][i] == self.board[2][i]):
                    return self.board[0][i]

        # diagonal wins
        if type(self.board[1][1]) is Player:
            if (self.board[0][0] == self.board[1][1] == self.board[2][2]) or \
                (self.board[0][2] == self.board[1][1] == self.board[2][0]):
                    return self.board[1][1]  

    def is_full(self):
        '''
        returns if board is filled
        '''

        for row in self.board:
            for cell in row:
                if cell == ' ':
                    return False
        return True

    def game_over(self):
        '''
        returns true if there is a winner or the board is full
        '''
        return self.is_full() or self.calc_winner()


if __name__ == '__main__':
    positions = {
        1: (0, 0), 2: (1, 0), 3: (2, 0),
        4: (0, 1), 5: (1, 1), 6: (2, 1),
        7: (0, 2), 8: (1, 2), 9: (2, 2)
    }

    game = TicTacToe()
    players = [
        Player(input('Player one: '), 'X'), 
        Player(input('Player two: '), 'O')
    ]
    
    current_round = 1
    print(game)

    while not game.game_over():
        player = players[current_round % 2]
        while True:
            try:
                move = int(input('Move: '))
                if not (0 < move < 10):
                    raise ValueError
                x, y = positions[move]
                if game.move(x, y, player):
                    print(game)
                    current_round += 1
                    break
                else: 
                    raise ValueError
            except ValueError:
                print('Enter an unoccupied position between 1-9.')

    winner = game.calc_winner()
    if winner:
        print(f'{winner.name} wins!')
    else:
        print(f'Tie!')

File: 3 JavaScript/solutions/vue-tic-tac-toe/tictactoe.js
// tictactoe.js vue implementation
Vue.component('cell', {
  props: ['position', 'i', 'addToken'],
  template:  `<div @click="addToken(i)">
                <span v-if='position'>{{position.token}}</span>
              </div>`
})


const app = new Vue({
  el: '#app',
  data: {
    message: 'Player 1\'s turn',
    players: [
      {name: 'Player 1', token: 'O', class:'one'},
      {name: 'Player 2', token: 'X', class:'two'}
    ],
    round: 1,
    board: Array.from(Array(9)), // empty array of length 9
  },
  methods: {
    addToken(idx) {
      this.message = `${this.player.name}'s turn`
      if (this.board[idx]) { 
      // position taken
        this.message = 'Invalid move'
      } else { 
      // move player to position and increment round
        this.board[idx] = this.player
        this.round++
        return this.gameOver()
      }
    },
    calcWinner() {
      for (let i=0; i<3; i++) {
        // horizontal wins
        const hcell = this.board[3*i]
        if (hcell && (this.board.slice(3*i, 3*i+3).filter(elem => elem === hcell).length === 3)) {
          console.log('horizontal', 3*i, 3*i+3, (this.board.slice(3*i, 3*i+3)))
          return hcell
        }
        // vertical wins
        const vcell = this.board[i]
        if (vcell && [this.board[i], this.board[i+3], this.board[i+6]].filter(elem => elem === vcell).length === 3) {
          console.log('vertical', i, i+3, i+6, ([this.board[i], this.board[i+3], this.board[i+6]]))
          return vcell
        }
      }
      // diagonal wins
      const cell = this.board[4] // center
      if (cell && this.board[0] === cell && cell === this.board[8]) return cell
      if (cell && this.board[2] === cell && cell === this.board[6]) return cell

      return null
    },
    boardFull() {
      for (let i=0; i<this.board.length; i++) {
        if (!this.board[i]) return false
      }
      return true
    },
    gameOver() {
      const winner = this.calcWinner()
      const message = (winner ? winner.name : false) || (this.boardFull() ? 'tie' : false)
      if (message) this.message = 'Winner: ' + message
      return message
    },
    reset() {
      this.board = Array.from(Array(9))
    }
  },
  computed: {
    player() {
      return this.players[this.round%2]
    },
  }
})